KAFKA-15931: Cancel RemoteLogReader gracefully

jeqo opened the pull request:
Reverts commit https://github.com/jeqo/kafka/commit/2723dbf3a007af0a1cafc17ed019aabd46eb6881 and  https://github.com/jeqo/kafka/commit/269e8892ad26904c82a3bff7b7d7a81c47619e4a.

Instead of reopening the transaction index, it cancels the RemoteFetchTask without interrupting it--avoiding to close the TransactionIndex channel.
This will lead to complete the exection of the remote fetch but ignoring the results. Given that this is considered a rare case, we could live with this. If it becomes a performance issue, it could be optimized.

jeqo committed:
Revert "MINOR: fix add missing @Test lost while rebasing (#19149)"

This reverts commit 2723dbf3a007af0a1cafc17ed019aabd46eb6881.

jeqo committed:
Revert "KAFKA-15931: Reopen TransactionIndex if channel is closed (#15241)"

This reverts commit 269e8892ad26904c82a3bff7b7d7a81c47619e4a.

jeqo committed:
KAFKA-15931: Cancel RemoteLogReader gracefully

Instead of interrupting the RemoteLogReader thread, use a cancel
operation to reply to the callback and set the thread as cancelled.
This should avoid getting the IO interrupted, leading to cached
TransactionIndex getting closed.

junrao commented on core/src/main/java/kafka/log/remote/RemoteLogReader.java:
```diff
@@ -57,8 +59,18 @@ public RemoteLogReader(RemoteStorageFetchInfo fetchInfo,
+
     @Override
     public Void call() {
+        if (cancelled) {
```
We probably want to pass `cancelled` to `rlm.read()` so that we could complete the task early if `cancelled` is true.

jeqo replied:
I'm failing to see how would passing `cancelled` to the `rlm.read()` work here. 

IIUC, when `rlm.read()` is running, it's already late, and we should let the task complete. By cancelling before `rlm.read()` is called here at `RemoteLogRead.call` it's the only place where it could complete early. Though I may be missing something

jeqo replied:
Another question: would it make sense to make RemoteLogReader extend ShutdownableThread? I just wonder if that was the initial suggestion when pointing to CleanerThread and other non-interruptible threads. I haven't applied this yet as those threads are usually started manually, instead of managing them via thread pools IIUC.

junrao replied:
When a task is cancelled, it could be in the middle of the following code in `rlm.read()`. I was thinking that if we keep checking the cancelled flag during the iteration, we could short-circuit it.

`                Utils.readFully(remoteSegInputStream, buffer);`

However, this is a rare situation and is probably an over optimization. For now, I think just cancelling the future is enough. We could remove the `cancelled` flag in this class.

> Another question: would it make sense to make RemoteLogReader extend ShutdownableThread? 
We can't easily used that. We will need to implement some sort of a task queue on top of it. It's simpler to just use the ExecutorService.

jeqo replied:
Got it. So the approach will be to just let that thread to finish and ignore the result, so the channel is not closed; right?

junrao replied:
Yes. When the task is cancelled, if it hasn't been scheduled, executor service will skip it. If it has, we can just let it finish and the result will be ignored. The latter is a bit wasteful, but is rare. We could optimize that later if it becomes a real problem.

junrao commented on core/src/main/scala/kafka/server/DelayedRemoteFetch.scala:
```diff
@@ -88,8 +90,9 @@ class DelayedRemoteFetch(remoteFetchTask: Future[Void],
     // cancel the remote storage read task, if it has not been executed yet
-    val cancelled = remoteFetchTask.cancel(true)
-    if (!cancelled) debug(s"Remote fetch task for RemoteStorageFetchInfo: $remoteFetchInfo could not be cancelled and its isDone value is ${remoteFetchTask.isDone}")
+    val cancelled = remoteFetchTask.cancel(false)
```
Hmm, the original logic doesn't look quite right. When a delayed task expires due to the timeout, we first call `onComplete()` and then `onExpiration()`. So cancelling the task in `onExpiration()` is too late since we call `remoteFetchResult.get` in `onComplete()`. 

To address this issue, we could call `remoteFetchTask.cancel()` in `onComplete()` if `remoteFetchResult.isDone` is false.

jeqo replied:
Aha, thanks for the clarification.

Agree, we could cancel the future, and use the `!future.isCancelled` for logging on the `onExpiration()` instead

junrao replied:
Could we add a comment on why it's important to cancel without interruption?

jeqo replied:
Sure! Adding it on the last commit.

junrao commented on core/src/main/java/kafka/log/remote/RemoteLogReader.java:
```diff
@@ -57,8 +59,18 @@ public RemoteLogReader(RemoteStorageFetchInfo fetchInfo,
 
+    public void cancel() {
+        LOGGER.debug("Cancelling remote log reader for topic partition {}", fetchInfo.topicPartition);
+        callback.accept(new RemoteLogReadResult(Optional.empty(), Optional.of(new InterruptedException("Cancelled remote log reader"))));
```
1. Hmm, doing this here would mean that we could call the callback twice?
2. Another problem is that the caller is from the DelayedRemoteFetch.onComplete(). The callback calls `delayedRemoteFetchPurgatory.checkAndComplete(key)`. Ideally, we don't want to call the purgatory while we are inside the purgatory to avoid potential deadlocks and infinite stacks.

jeqo replied:
Thanks for pointing this out. I'm moving this callback call out of the `cancel()`

junrao commented on core/src/main/scala/kafka/server/DelayedRemoteFetch.scala:
```diff
@@ -36,6 +37,7 @@ import scala.collection._
  * in the remote fetch operation purgatory
  */
 class DelayedRemoteFetch(remoteFetchTask: Future[Void],
+                         remoteLogReader: RemoteLogReader,
```
This is an existing issue. It would be better if we could consolidate remoteFetchTask and remoteFetchResult into a single future.

jeqo replied:
Good idea! I'm giving a try to this by defining a `RemoteFetchFuture` that includes both futures and the RemoteLogReader to handle all the cancel procedure.

junrao reviewed (commented):
@jeqo : Thanks for the PR. Left a few comments.

jeqo committed:
apply suggestions

junrao commented on core/src/main/scala/kafka/server/DelayedRemoteFetch.scala:
```diff
@@ -119,6 +118,9 @@ class DelayedRemoteFetch(remoteFetchTask: Future[Void],
       } else {
+        // cancel the remote storage read task, if it has not been executed yet
+        if (!remoteFetchFuture.isDone)
+          remoteFetchFuture.cancel()
```
Sorry, I didn't realize this branch doesn't depend on the result from the future. So, if the future is not done, we simply return empty. Given this, it's better to call remoteFetchFuture.cancel() in onExpiration() as it was before.

jeqo replied:
Right. Moving this cancel back to the onExpiration

junrao commented on core/src/main/scala/kafka/server/ReplicaManager.scala:
```diff
@@ -1544,12 +1544,14 @@ class ReplicaManager(val config: KafkaConfig,
         delayedRemoteFetchPurgatory.checkAndComplete(key)
       })
+      val remoteFetchTask = remoteLogManager.get.asyncRead(remoteLogReader)
+      remoteFetchFuture = new RemoteFetchFuture(remoteLogReader, remoteFetchTask)
```
Hmm, is this correct? We need to pass in remoteFetchFuture to RemoteFetchFuture, right?

jeqo replied:
I think you mean the result future right? If yes; agree, I missed it.
I'm passing both future as input now.

junrao reviewed (commented):
@jeqo : Thanks for the updated PR. A few more comments.

jeqo committed:
move cancel back to expiration

jeqo committed:
removing cancelled flag from RLR and RLR from RemoteFetchFuture

cancelling RLR seems like an overoptimization atm, then moving it out of
the RemoteFetchFuture

jeqo committed:
pass result future to RemoteFetchFuture

junrao commented on core/src/main/scala/kafka/server/ReplicaManager.scala:
```diff
@@ -1558,7 +1559,7 @@ class ReplicaManager(val config: KafkaConfig,
 
     val remoteFetchMaxWaitMs = config.remoteLogManagerConfig.remoteFetchMaxWaitMs().toLong
-    val remoteFetch = new DelayedRemoteFetch(remoteFetchTask, remoteFetchResult, remoteFetchInfo, remoteFetchMaxWaitMs,
+    val remoteFetch = new DelayedRemoteFetch(remoteFetchFuture, remoteFetchInfo, remoteFetchMaxWaitMs,
```
Since we only need to pass in `remoteFetchTask` and `remoteFetchResult`, it seems that it's simpler to just pass them in separately as before, instead of wrapping them under `RemoteFetchFuture`.

jeqo replied:
Right. In that case the main change would be cancelling the future without interrupting it.

junrao commented on core/src/main/scala/kafka/server/DelayedRemoteFetch.scala:
```diff
@@ -18,6 +18,7 @@
 package kafka.server
 
 import com.yammer.metrics.core.Meter
+import kafka.log.remote.RemoteLogReader
```
unused import

junrao reviewed (commented):
@jeqo : Thanks for the updated PR. A couple of more comments.

jeqo committed:
reformats

jeqo committed:
return to two futures

junrao reviewed (commented):
@jeqo : Thanks for the updated PR. Just a minor comment. Could you take the PR out of the draft mode?

jeqo committed:
add comment on why not interrupting task

junrao reviewed (approved):
@jeqo : Thanks for the updated PR. LGTM
